src/machine/machine_avr.go

```diff
@@ -58,3 +58,93 @@ func (p GPIO) Get() bool {
 		return (val > 0)
 	}
 }
+
+const (
+	COM2A1 = 7
+	COM2A0 = 6
+	COM2B1 = 5
+	COM2B0 = 4
+	COM0A1 = 7
+	COM0A0 = 6
+	COM0B1 = 5
+	COM0B0 = 4
+	COM1A1 = 7
+	COM1A0 = 6
+	COM1B1 = 5
+	COM1B0 = 4
+
+	CS00 = 0
+	CS01 = 1
+	CS11 = 1
+	CS22 = 2
+
+	WGM10 = 0
+	WGM11 = 1
+	WGM20 = 0
+	WGM21 = 1
+)
+
+// InitPWM initializes the registers needed for PWM.
+func InitPWM() {
+	// use waveform generation
+	*avr.TCCR0A |= avr.TCCR0A_WGM0
+
+	// set timer 0 prescale factor to 64
+	*avr.TCCR0B |= 1 << CS01
+	*avr.TCCR0B |= 1 << CS00
+
+	// set timer 1 prescale factor to 64
+	*avr.TCCR1B |= 1 << CS11
+
+	// put timer 1 in 8-bit phase correct pwm mode
+	*avr.TCCR1A |= 1 << WGM10
+
+	// set timer 2 prescale factor to 64
+	*avr.TCCR2B |= 1 << CS22
+
+	// configure timer 2 for phase correct pwm (8-bit)
+	*avr.TCCR2A |= 1 << WGM20
+}
+
+// Configure configures a PWM pin for output.
+func (pwm PWM) Configure() {
+	if pwm.Pin < 8 {
+		*avr.DDRD |= 1 << pwm.Pin
+	} else {
+		*avr.DDRB |= 1 << (pwm.Pin - 8)
+	}
+}
+
+// Set sets the needed register values to turn on the duty cycle for a PWM pin.
+func (pwm PWM) Set(value uint8) {
+	switch pwm.Pin {
+	case 3:
+		// connect pwm to pin on timer 2, channel B
+		*avr.TCCR2A |= 1 << COM2B1
+		*avr.OCR2B = avr.RegValue(value) // set pwm duty
+	case 5:
+		// connect pwm to pin on timer 0, channel B
+		*avr.TCCR0A |= 1 << COM0B1
+		*avr.OCR0B = avr.RegValue(value) // set pwm duty
+	case 6:
+		// connect pwm to pin on timer 0, channel A
+		*avr.TCCR0A |= 1 << COM0A1
+		*avr.OCR0A = avr.RegValue(value) // set pwm duty
+	case 9:
+		// connect pwm to pin on timer 1, channel A
+		*avr.TCCR1A |= 1 << COM1A1
+		// this is a 16-bit value, but we only currently allow the low order bits to be set
+		*avr.OCR1AL = avr.RegValue(value) // set pwm duty
+	case 10:
+		// connect pwm to pin on timer 1, channel B
+		*avr.TCCR1A |= 1 << COM1B1
+		// this is a 16-bit value, but we only currently allow the low order bits to be set
+		*avr.OCR1BL = avr.RegValue(value) // set pwm duty
+	case 11:
+		// connect pwm to pin on timer 2, channel A
+		*avr.TCCR2A |= 1 << COM2A1
+		*avr.OCR2A = avr.RegValue(value) // set pwm duty
+	default:
+		// TODO: handle invalid pin for PWM on Arduino
```

Comment:
I think it's best to specify what should happen here, and document it at the function documentation. Either not doing anything, or panicking (both is fine to me). Panicking should be fine in case the pin number is known at compile time, as this whole switch will be optimized away and thus also the call to `panic()`.

Reply:
In that case a panic() seems like it might be the best here for compile time check. 

How should it handle the cases someone tries to specify at runtime which pin to use? For example to write a Firmata-like implementation that is controlled by an external caller using serial interface, or wireless interface or something.

Reply:
When it is specified at runtime (or LLVM can't deduce it - it isn't magical) it will be a regular switch, like before. And that means the panic will be compiled in the binary, which you may or may not want. I don't have a strong preference for either: having a panic here makes finding bugs easier in some cases but may also increase firmware size.